extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/sync/CatalogUpdateNotifier.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.druid.discovery.DruidNodeDiscoveryProvider;
+import org.apache.druid.discovery.NodeRole;
+import org.apache.druid.guice.ManageLifecycle;
+import org.apache.druid.guice.annotations.EscalatedClient;
+import org.apache.druid.guice.annotations.Smile;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStart;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStop;
+import org.apache.druid.java.util.http.client.HttpClient;
+import org.apache.druid.server.DruidNode;
+import org.joda.time.Duration;
+
+import javax.inject.Inject;
+
+import java.util.Collections;
+import java.util.function.Supplier;
+
+/**
+ * Global update notifier for the catalog. Registers itself as a catalog
+ * listener, then uses the common cache notifier to send Smile-encoded JSON
+ * updates to broker nodes discovered from node discovery (typically ZooKeeper.)
+ * <p>
+ * Deletes are encoded as a table update with a table definition of a special
+ * tombstone type. This saves having the need for two endpoints, or having
+ * a wrapper class to handle deletes.
+ */
+@ManageLifecycle
+public class CatalogUpdateNotifier implements CatalogListener
+{
+  private static final String CALLER_NAME = "Catalog Sync";
+  private static final long TIMEOUT_MS = 5000;
+
+  /**
+   * Internal table type used in updates to notify listeners that a table has
+   * been deleted. Avoids the need for a special "table deleted" message.
+   */
+  public static final String TOMBSTONE_TABLE_TYPE = "tombstone";
+  private static final TableSpec TABLE_TOMBSTONE = new TableSpec(TOMBSTONE_TABLE_TYPE, null, null);
+
+  private final CacheNotifier notifier;
+  private final ObjectMapper smileMapper;
+
+  @Inject
+  public CatalogUpdateNotifier(
+      CatalogStorage catalog,
+      DruidNodeDiscoveryProvider discoveryProvider,
+      @EscalatedClient HttpClient httpClient,
+      @Smile ObjectMapper smileMapper
+  )
+  {
+    long timeoutMs = TIMEOUT_MS;
+    this.smileMapper = smileMapper;
+    Supplier<Iterable<DruidNode>> nodeSupplier = new ListeningNodeSupplier(
+        Collections.singletonList(NodeRole.BROKER),
+        discoveryProvider);
+    RestSender restSender = RestUpdateSender.httpClientSender(httpClient, Duration.millis(timeoutMs));
+    RestUpdateSender sender = new RestUpdateSender(
+        CALLER_NAME,
+        nodeSupplier,
+        restSender,
+        CatalogListenerResource.BASE_URL + CatalogListenerResource.SYNC_URL,
+        timeoutMs);
+    this.notifier = new CacheNotifier(
+        CALLER_NAME,
+        sender);
+    catalog.register(this);
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    notifier.start();
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    notifier.stop();
+  }
+
+  @Override
+  public void updated(TableMetadata update)
+  {
+    notifier.send(update.toBytes(smileMapper));
+  }
+
+  @Override
+  public void deleted(TableId tableId)
+  {
+    TableMetadata spec = TableMetadata.newTable(tableId, TABLE_TOMBSTONE);
```

Comment:
If the listener wanted to delegate proper deletion based on the type of the table, how would they do this with this special tombstone type? For example say some type of table's deletion process is different than another.

Reply:
The notification mechanism has changed to use events, so no more tombstones. The deletion event here is that the table was deleted from the catalog DB. This can happen because the user made a mistake, and just deletes an unwanted entry. Or, more formally, it would be the last step in a deletion process outlined in an earlier comment. This event should trigger nothing other than removal from a cache on the broker.